python/dgl/homophily.py

```diff
@@ -45,13 +59,140 @@ def node_homophily(graph, y):
+        # Handle the case where graph is of dtype int32.
+        src, dst = get_long_edges(graph)
+        # Compute y_v = y_u for all edges.
+        edge_indicator = F.astype(y[src] == y[dst], F.float32)
+        return F.as_scalar(F.mean(edge_indicator, dim=0))
+
+
+def linkx_homophily(graph, y):
+    r"""Homophily measure from `Large Scale Learning on Non-Homophilous Graphs:
+    New Benchmarks and Strong Simple Methods
+    <https://arxiv.org/abs/2110.14446>`__
+
+    Mathematically it is defined as follows:
+
+    .. math::
+      \frac{1}{C-1} \sum_{k=1}^{C} \max \left(0, \frac{\sum_{v\in C_k}|\{u\in
+      \mathcal{N}(v): y_v = y_u \}|}{\sum_{v\in C_k}|\mathcal{N}(v)|} -
+      \frac{|\mathcal{C}_k|}{|\mathcal{V}|} \right)
+
+    where :math:`C` is the number of node classes, :math:`C_k` is the set of
+    nodes that belong to class k, :math:`\mathcal{N}(v)` are the predecessors
+    of node :math:`v`, :math:`y_v` is the class of node :math:`v`, and
+    :math:`\mathcal{V}` is the set of nodes.
+
+    Parameters
+    ----------
+    graph : DGLGraph
+        The graph
+    y : Tensor
+        The node labels, which is a tensor of shape (|V|)
+
+    Returns
+    -------
+    float
+        The homophily value
+
+    Examples
+    --------
+    >>> import dgl
+    >>> import torch
+
+    >>> graph = dgl.graph(([0, 1, 2, 3], [1, 2, 0, 4]))
+    >>> y = torch.tensor([0, 0, 0, 0, 1])
+    >>> dgl.linkx_homophily(graph, y)
+    0.19999998807907104
+    """
+    with graph.local_scope():
+        # Compute |{u\in N(v): y_v = y_u}| for each node v.
+        # Handle the case where graph is of dtype int32.
+        src, dst = get_long_edges(graph)
+        # Compute y_v = y_u for all edges.
+        graph.edata["same_class"] = (y[src] == y[dst]).float()
+        graph.update_all(
+            fn.copy_e("same_class", "m"), fn.mean("m", "same_class_deg")
+        )
+
+        # Compute |N(v)| for each node v.
+        deg = graph.in_degrees().float()
+
+        # To compute \sum_{v\in C_k}|{u\in N(v): y_v = y_u}| for all k
+        # efficiently, construct a directed graph from nodes to their class.
+        num_classes = F.max(y, dim=0).item() + 1
+        src = graph.nodes().to(dtype=y.dtype)
+        dst = y + graph.num_nodes()
+        class_graph = create_graph((src, dst))
+        # Add placeholder values for the class nodes.
+        class_placeholder = torch.zeros(
+            (num_classes), dtype=deg.dtype, device=class_graph.device
+        )
+        class_graph.ndata["same_class_deg"] = torch.cat(
+            [graph.ndata["same_class_deg"], class_placeholder], dim=0
+        )
+        class_graph.update_all(
+            fn.copy_u("same_class_deg", "m"), fn.sum("m", "class_deg_aggr")
+        )
+
+        # Similarly, compute \sum_{v\in C_k}|N(v)| for all k in parallel.
+        class_graph.ndata["deg"] = torch.cat([deg, class_placeholder], dim=0)
+        class_graph.update_all(fn.copy_u("deg", "m"), fn.sum("m", "deg_aggr"))
```

Comment:
I feel the implementation is over complicated and creating a new graph maybe costly too when the input graph is large. Will the following be simpler/more-readable?

1. Compute the node heterophily without averaging it across all the nodes.
2. Have a loop over the number of classes and each iteration slices the nodes of one class and compute the class-wise heterophily rate.
3. Aggregate across all the classes.

As the number of classes is usually small, the loop won't be too slow.

Reply:
simplified